File: cracking_the_coding_interview_6th/chapter16/cpp/16_26_v02.cpp
// Cracking the Coding Interview p.525
#include <fstream>
#include <regex>
#include <string>
#include <vector>
using namespace std;
using ll = long long;

class Term {
public:
  enum class Operator { ADD, SUB, MUL, DIV, BLANK };
  double v;
  Operator op;
  Term(double _v, Operator _op) : v(_v), op(_op) {}
  static vector<Term> parse_term_seq(string);
  const static regex TERM_PAT;
  string repr() {
    string res;
    switch (op) {
    case Operator::ADD:
      res += "+";
      break;
    case Operator::SUB:
      res += "-";
      break;
    case Operator::MUL:
      res += "*";
      break;
    case Operator::DIV:
      res += "/";
      break;
    default:
      res += " ";
      break;
    }
    return res += to_string(v);
  }
};

vector<Term> Term::parse_term_seq(string seq) {
  vector<Term> res;
  regex pat{R"(([\+\-\*\/]?)([0-9]+))"};
  for (sregex_iterator p(seq.begin(), seq.end(), pat); p != sregex_iterator{};
       ++p) {
    if ((*p)[1] == '+')
      res.emplace_back(stod((*p)[2]), Operator::ADD);
    else if ((*p)[1] == '-')
      res.emplace_back(stod((*p)[2]), Operator::SUB);
    else if ((*p)[1] == '*')
      res.emplace_back(stod((*p)[2]), Operator::MUL);
    else if ((*p)[1] == '/')
      res.emplace_back(stod((*p)[2]), Operator::DIV);
    else
      res.emplace_back(stod((*p)[2]), Operator::BLANK);
  }
  return res;
}

double apply_op(double left, Term::Operator op, double right) {
  switch (op) {
  case Term::Operator::ADD:
    return left + right;
  case Term::Operator::SUB:
    return left - right;
  case Term::Operator::MUL:
    return left * right;
  case Term::Operator::DIV:
    return left / right;
  default:
    return right;
  }
}

Term *collapse_term(Term *primary, Term *secondary) {
  if (!primary) return secondary;
  if (!secondary) return primary;
  double v = apply_op(primary->v, secondary->op, secondary->v);
  primary->v = v;
  return primary;
}

double compute(string seq) {
  vector<Term> terms = Term::parse_term_seq(seq);
  double res = 0;
  Term *processing = nullptr;
  for (ll i = 0; i < terms.size(); ++i) {
    Term *curr = &terms[i];
    Term *nxt = i + 1 < terms.size() ? &terms[i + 1] : nullptr;
    processing = collapse_term(processing, curr);
    if (nxt == nullptr || nxt->op == Term::Operator::ADD
        || nxt->op == Term::Operator::SUB) {
      res = apply_op(res, processing->op, processing->v);
      processing = nullptr;
    }
  }
  return res;
}

void solve(const string &fp) {
  printf("# %s\n", fp.c_str());
  ifstream ifs{fp};
  string expr;
  ifs >> expr;
  printf("%f\n", compute(expr));
  puts("");
}

int main() {
  solve("../testcases/16_26/01.txt");
  solve("../testcases/16_26/02.txt");
  solve("../testcases/16_26/03.txt");
}

// # ../testcases/16_26/01.txt
// 23.500000

// # ../testcases/16_26/02.txt
// 120.000000

// # ../testcases/16_26/03.txt
// 99.000000

File: cracking_the_coding_interview_6th/chapter16/python/16_26_v02.py
from __future__ import annotations
from typing import List, Union
from enum import Enum, auto
import re


class Term:
    class Operator(Enum):
        ADD = auto()
        SUB = auto()
        MUL = auto()
        DIV = auto()
        BLANK = auto()

    @classmethod
    def parse_term_seq(cls, seq: str) -> List[Term]:
        res = []
        pat = re.compile(r"([\+\-\*\/]?)([0-9]+)")
        for m in pat.finditer(seq):
            if m[1] == '+':
                res.append(Term(float(m[2]), cls.Operator.ADD))
            elif m[1] == '-':
                res.append(Term(float(m[2]), cls.Operator.SUB))
            elif m[1] == '*':
                res.append(Term(float(m[2]), cls.Operator.MUL))
            elif m[1] == '/':
                res.append(Term(float(m[2]), cls.Operator.DIV))
            else:
                res.append(Term(float(m[2]), cls.Operator.BLANK))
        return res

    def __init__(self, v: float, op: Operator):
        self.v: float = v
        self.op: self.Operator = op


def apply_op(left: float, op: Term.Operator, right: float) -> float:
    if op is Term.Operator.ADD:
        return left + right
    elif op is Term.Operator.SUB:
        return left - right
    elif op is Term.Operator.MUL:
        return left * right
    elif op is Term.Operator.DIV:
        return left / right
    else:
        return right


def collapse_term(primary: Union[Term, None], secondary: Union[Term, None]) -> Term:
    if primary is None:
        return secondary
    if secondary is None:
        return primary
    v = apply_op(primary.v, secondary.op, secondary.v)
    primary.v = v
    return primary


def compute(seq: str) -> float:
    terms = Term.parse_term_seq(seq)
    res = 0
    processing = None
    for i in range(len(terms)):
        curr = terms[i]
        nxt = terms[i + 1] if i + 1 < len(terms) else None
        processing = collapse_term(processing, curr)
        if nxt is None or nxt.op is Term.Operator.ADD or nxt.op is Term.Operator.SUB:
            res = apply_op(res, processing.op, processing.v)
            processing = None
    return res


def ns(f):
    return next(f).strip()


def solve(fp: str):
    print(f"# {fp}")
    with open(fp) as f:
        expr = ns(f)
    print(compute(expr))
    print()


solve("../testcases/16_26/01.txt")
solve("../testcases/16_26/02.txt")
solve("../testcases/16_26/03.txt")

# # ../testcases/16_26/01.txt
# 23.5

# # ../testcases/16_26/02.txt
# 120.0

# # ../testcases/16_26/03.txt
# 99.0
